digdag-standards/src/main/java/io/digdag/standards/operator/td/TdRunOperatorFactory.java

```diff
@@ -42,44 +53,84 @@ public Operator newTaskExecutor(Path workspacePath, TaskRequest request)
+        private final String name;
+        private final Instant sessionTime;
+        private final Optional<String> downloadFile;
+        private final boolean storeLastResults;
+        private final boolean preview;
+        private final Config state;
+        private final Optional<String> existingJobId;
+        private final Optional<String> existingDomainKey;
+
         public TdRunOperator(Path workspacePath, TaskRequest request)
         {
             super(workspacePath, request);
+
+            this.params = request.getConfig().mergeDefault(
+                    request.getConfig().getNestedOrGetEmpty("td"));
+
+            this.name = params.get("_command", String.class);
+            this.sessionTime = params.get("session_time", Instant.class);
+            this.downloadFile = params.getOptional("download_file", String.class);
+            this.storeLastResults = params.get("store_last_results", boolean.class, false);
+            this.preview = params.get("preview", boolean.class, false);
+
+            this.state = request.getLastStateParams().deepCopy();
+
+            this.existingJobId = state.getOptional(JOB_ID, String.class);
+            this.existingDomainKey = state.getOptional(DOMAIN_KEY, String.class);
         }
 
         @Override
         public TaskResult runTask()
         {
-            Config params = request.getConfig().mergeDefault(
-                    request.getConfig().getNestedOrGetEmpty("td"));
+            try (TDOperator op = TDOperator.fromConfig(params)) {
 
-            String name = params.get("_command", String.class);
-            Instant sessionTime = params.get("session_time", Instant.class);
-            Optional<String> downloadFile = params.getOptional("download_file", String.class);
-            boolean storeLastResults = params.get("store_last_results", boolean.class, false);
-            boolean preview = params.get("preview", boolean.class, false);
+                // Generate and store domain key before starting the job
+                if (!existingDomainKey.isPresent()) {
+                    String domainKey = UUID.randomUUID().toString();
+                    state.set(DOMAIN_KEY, domainKey);
+                    throw TaskExecutionException.ofNextPolling(0, ConfigElement.copyOf(state));
+                }
 
-            try (TDOperator op = TDOperator.fromConfig(params)) {
-                TDJobOperator j = op.startSavedQuery(name, Date.from(sessionTime));
-                logger.info("Started a saved query name={} with time={}", name, sessionTime);
+                // Start the job
+                if (!existingJobId.isPresent()) {
+                    TDJobOperator j = op.startSavedQuery(name, Date.from(sessionTime), existingDomainKey.get());
+                    logger.info("Started a saved query name={} with time={}", name, sessionTime);
+                    state.set(JOB_ID, j.getJobId());
+                    state.set(POLL_ITERATION, 1);
```

Comment:
Yeah, I want to do that in a separate PR that factors out all the duplicated non-blocking job execution logic from the td operators into reusable code.